lib/Transforms/LLZKR1CSLoweringPass.cpp

```diff
@@ -0,0 +1,314 @@
+private:
+  unsigned auxCounter = 0;
+
+  // Normalize a felt-valued expression into R1CS-compatible form.
+  // This performs *minimal* rewriting:
+  // - Only rewrites Add/Sub of two degree-2 terms
+  // - Operates bottom-up using post-order traversal
+  //
+  // Resulting expression is R1CS-compatible (i.e., one multiplication per constraint)
+  // and can be directly used in EmitEqualityOp or as operands of other expressions.
+
+  void getPostOrder(Value root, SmallVectorImpl<Value> &postOrder) {
+    SmallVector<Value, 16> worklist;
+    DenseSet<Value> visited;
+
+    worklist.push_back(root);
+
+    while (!worklist.empty()) {
+      Value val = worklist.back();
+
+      if (visited.contains(val)) {
+        worklist.pop_back();
+        postOrder.push_back(val);
+        continue;
+      }
+
+      visited.insert(val);
+      if (Operation *op = val.getDefiningOp()) {
+        for (Value operand : op->getOperands()) {
+          worklist.push_back(operand);
+        }
+      }
+    }
+  }
+
+  /// Normalize a felt-valued expression into R1CS-compatible form by rewriting
+  /// only when strictly necessary. This function ensures the resulting expression:
+  ///
+  /// - Has at most one multiplication per constraint (R1CS-compatible)
+  /// - Avoids unnecessary introduction of auxiliary variables
+  /// - Preserves semantic equivalence via auxiliary field equality constraints
+  ///
+  /// Rewriting is done **bottom-up** using post-order traversal of the def-use chain.
+  /// The transformation is minimal:
+  /// - Only rewrites Add/Sub where both operands are degree-2
+  /// - Leaves multiplications intact unless their operands require rewriting due to constants
+  /// - Avoids rewriting expressions that are already linear or already normalized
+  ///
+  /// The function memoizes all degrees and rewrites for efficiency and correctness,
+  /// and records any auxiliary field assignments for later reconstruction in compute().
+  ///
+  /// \param root           The root felt-valued expression to normalize.
+  /// \param structDef      The enclosing struct definition (for adding aux fields).
+  /// \param constrainFunc  The constrain() function containing the constraint logic.
+  /// \param degreeMemo     Memoized degrees of expressions (to avoid recomputation).
+  /// \param rewrites       Memoized rewrites of expressions.
+  /// \param auxAssignments Records auxiliary field assignments introduced during normalization.
+  /// \param builder        Builder used to insert new ops in the constrain() block.
+  /// \returns              A Value representing the normalized (possibly rewritten) expression.
+  Value normalizeForR1CS(
+      Value root, StructDefOp structDef, FuncDefOp constrainFunc,
+      DenseMap<Value, unsigned> &degreeMemo, DenseMap<Value, Value> &rewrites,
+      SmallVectorImpl<AuxAssignment> &auxAssignments, OpBuilder &builder
+  ) {
+    if (rewrites.count(root)) {
+      return rewrites[root];
+    }
```

Comment:
```suggestion
    if (auto it = rewrites.find(root); it != rewrites.end()) {
      return it->second;
    }
```

Reply:
Maybe just `rewrite.contains(root)`?

Reply:
What you have now is fine. However, since you are using the value in this case, the `find()` approach avoids double lookup.

Reply:
Ok cool. I'll change it to what you suggest.